source/dump.f90

```diff
@@ -78,37 +67,121 @@ module dump
-  call alloc(dumptasinv,            nblz_new, 6, 6, zero, "dumptasinv", -1,1,1)
-  call alloc(dumpclo,               nblz_new, 6,    zero, "dumpclo",    -1,1)
+  call alloc(ldump,                 nele_new, .false., "ldump",       -1)
+  call alloc(ndumpt,                nele_new, 0,       "ndumpt",      -1)
+  call alloc(dumpfirst,             nele_new, 0,       "dumpfirst",   -1)
+  call alloc(dumplast,              nele_new, 0,       "dumplast",    -1)
+  call alloc(dumpunit,              nele_new, 0,       "dumpunit",    -1)
+  call alloc(dumpfmt,               nele_new, 0,       "dumpfmt",     -1)
+  call alloc(dump_fname, mFileName, nele_new, " ",     "dump_fname",  -1)
+  call alloc(dump_nMatMap,          nele_new, 0,       "dump_nMatMap",-1)
 
 #ifdef CR
-  call alloc(dumpfilepos,           nele_new,-1,          "dumpfilepos",   -1)
-  call alloc(dumpfilepos_cr,        nele_new,-1,          "dumpfilepos_cr",-1)
+  call alloc(dumpfilepos,           nele_new,-1,       "dumpfilepos",   -1)
+  call alloc(dumpfilepos_cr,        nele_new,-1,       "dumpfilepos_cr",-1)
 #endif
 
 #ifdef HDF5
-  call alloc(dump_hdf5DataSet,      nele_new,0,           "dump_hdf5DataSet",-1)
-  call alloc(dump_hdf5Format,       9,       0,           "dump_hdf5Format")
+  call alloc(dump_hdf5DataSet,      nele_new,0,        "dump_hdf5DataSet",-1)
+  call alloc(dump_hdf5Format,       9,       0,        "dump_hdf5Format")
 #endif
 
 end subroutine dump_expand_arrays
 
+! ================================================================================================ !
+!  V.K. Berglyd Olsen
+!  Created: 2019-08-05
+!  Updated: 2019-08-05
+!  Set the normalisation matrix, compute its inverse, and save it in a compact array of structs.
+! ================================================================================================ !
+subroutine dump_setTasMatrix(elemID, tasData, cloData)
+
+  use crcoall
+  use mod_common, only : bez
+  use mod_settings
+
+  integer,          intent(in) :: elemID
+  real(kind=fPrec), intent(in) :: tasData(6,6)
+  real(kind=fPrec), intent(in) :: cloData(6)
+
+  type(dump_normType), allocatable :: tmpNorm(:)
+  real(kind=fPrec) invData(6,6)
+  integer storeID
+
+  if(dump_nMatMap(elemID) == 0) then
+    if(allocated(dump_normMat)) then
+      allocate(tmpNorm(dump_nNormMat+1))
+      tmpNorm(1:dump_nNormMat) = dump_normMat(1:dump_nNormMat)
+      call move_alloc(tmpNorm, dump_normMat)
+      dump_nNormMat = dump_nNormMat + 1
+    else
+      allocate(dump_normMat(1))
+      dump_nNormMat = 1
+    end if
+    dump_nMatMap(elemID) = dump_nNormMat
+    storeID = dump_nNormMat
+    if(st_debug) then
+      if(elemID > 0) then
+        write(lout,"(a,i0,a)") "DUMP> Saving normalisation matrix for element ",elemID," '"//trim(bez(elemID))//"'"
+      else if(elemID == -1) then
+        write(lout,"(a)") "DUMP> Saving normalisation matrix for element 'StartDUMP'"
+      end if
+    end if
+  else
```

Comment:
Why is this even allowed? Seems like a bug (uncaught input error) if it happens - the results won't be valid...

Reply:
I added it originally because the setting of the matrix was called three times for each element due to being in an existing loop with the other code. Now that I deleted the old code, the call is outside the loop in a different location than before.

In any case, the old code did not prevent the overwriting of the values at saving time, so I just replicated the behaviour for compact storage. I was trying to to impose new restrictions on the existing logic because it would trigger more work in rewriting DUMP.

Reply:
As far as I can see, there is nothing that checks that the element used for normalised dump is unique. We should probably require that for normalised dumps. Not sure where is the best place to add that. Adding it here also means we have to check for dump format, otherwise we will block all non-unique elements.

Reply:
Yeah, that's not really ideal in DUMP. At least an almost-bug :)

Maybe it's OK to just let it crash here then, with a suggestive error message?

Reply:
We cannot restrict this it turns out. The routine where this is saved from is called during 6D closed orbit search, so potentially multiple times. The functionality to allow updating the values need to stay until we look into this in more detail. Otherwise it breaks existing functionality.